lib/pbio/drv/sound/sound_ev3.c

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025 The Pybricks Authors
+
+#include <pbdrv/config.h>
+
+#if PBDRV_CONFIG_SOUND_EV3
+
+#include <stdint.h>
+
+#include <pbdrv/gpio.h>
+
+#include <tiam1808/ehrpwm.h>
+#include <tiam1808/hw/soc_AM1808.h>
+#include <tiam1808/hw/hw_syscfg0_AM1808.h>
+#include <tiam1808/hw/hw_types.h>
+#include <tiam1808/psc.h>
+
+#include "../drv/gpio/gpio_ev3.h"
+
+// This module covers the frequency range from 64 Hz to 10 kHz.
+// It uses a maximum duty cycle of 1/16, and it controls volume
+// by shortening the duty cycle. The hardware has a resolution
+// of 16 bits for the period counter, and we want 8 bits for the volume,
+// so we use the following timebase division factors to make this work:
+//
+// 64 Hz - 900 Hz   => /40
+// 900 Hz - 8 kHz   => /4
+// 8 kHz - 10 kHz   => /1
+
+// Audio amplifier enable
+static const pbdrv_gpio_t pin_sound_en = PBDRV_GPIO_EV3_PIN(13, 3, 0, 6, 15);
+// Audio output pin
+#define SYSCFG_PINMUX3_PINMUX3_7_4_GPIO0_0 0
+static const pbdrv_gpio_t pin_audio = PBDRV_GPIO_EV3_PIN(3, 7, 4, 0, 0);
+
+void pbdrv_sound_stop() {
+    // Turn speaker amplifier off
+    pbdrv_gpio_out_low(&pin_sound_en);
```

Comment:
There is a popping noise from the speaker every time we turn off the amplifier. So we should probably either leave it on all of the time or only turn it off after some timeout period. Otherwise, when we do beeps, there is a pop after every beep.

Or maybe there is some way we can program the PWM, e.g. 50kHz 50% duty cycle to get "0V" or invert the PWM output so that 0% duty cycle is high instead of low, that will minimize the popping sound?

Reply:
Listening carefully, it's not from turning *off* the amplifier, it's from turning *on* the amplifier.

*sigh* classic audio problems. We should probably indeed just leave it on and program the PWM controller to mute instead.